docs/minikube-walkthrough.md

```diff
@@ -0,0 +1,117 @@
+
+Minikube is a single node Kubernetes cluster that runs on your local machine. The Minikube docs have installation instructions for your OS. 
+- [minikube installation](https://github.com/kubernetes/minikube#installation)
+
+
+## Clone the repository
+
+In order to build the sample apiserver image we will need to build the apiserver binary. 
+
+```
+cd $GOPATH/src
+mkdir -p github.com/kubernetes
+cd github.com/kubernetes
+git clone https://github.com/kubernetes/sample-apiserver.git
+``` 
+
+## Rename the package imports
+
+After cloning from the repo, the imports will be referencing ```k8s.io/sample-apiserver``` these need to be changed to ```github.com/kubernetes/sample-apiserver```. Recommend you do this with your favourite editor and a find all and replace.
+
+## Build the binary
+
+Next we will want to create a new binary to both test we can build the server and to use for the container image.
+
+From the root of this repo, where ```main.go``` is located, run the following command:
+```
+export GOOS=linux; go build .
+```
+if everything went well, you should have a binary called ```sample-apiserver``` present in your current directory.
+
+## Build the container image
+
+Using the binary we just built, we will now create a Docker image and push it to our Dockerhub registry so that we deploy it to our cluster.
+There is a sample ```Dockerfile``` located in ```artifacts/simple-image``` we will use this to build our own image.
+
+Again from the root of this repo run the following commands:
+```
+cp ./sample-apiserver ./artifacts/simple-image/kube-sample-apiserver
+docker build -t <YOUR_DOCKERHUB_USER>/kube-sample-apiserver:latest ./artifacts/simple-image
+docker push <YOUR_DOCKERHUB_USER>/kube-sample-apiserver 
+```
+
+## Modify the the replication controller start command
+
+In the current version of Minikube ```v0.22.0``` at the time of writing, there is some missing configuration required for running an extension apiserver. This has been fixed in the following [PR](https://github.com/kubernetes/minikube/pull/1918) so hopefully in the next release this step will no longer be needed:
+
+In your favourite editor open the replication controller template:
+
+``` 
+vi artifacts/example/rc.yaml
+```
+Under the containers sections, right after the line that specifies the etcd server address : ```- "--etcd-servers=http://localhost:2379"``` add the following:
+
+```
+        - --requestheader-client-ca-file=/var/run/secrets/kubernetes.io/serviceaccount/ca.crt
+        - --requestheader-username-headers=X-Remote-User
+        - --requestheader-group-headers=X-Remote-Group
+        - --requestheader-extra-headers-prefix=X-Remote-Extra        
+```
+
+You need to change the ```imagePullPolicy``` to ```Always``` or ```IfNotPresent```.
+
+You also need to change the image from ```kube-sample-apiserver:latest``` to ```<YOUR_DOCKERHUB_USER>/kube-sample-apiserver:latest```
+
+Save this file and we are then ready to deploy and try out the sample apiserver.
```

Comment:
I wonder whether we shouldn't deliver artifacts for this use-case and avoid instructions like this to modify files. Would make it much easier.

Reply:
Not sure I am following. Do you mean provide a image somewhere like dockerhub? That would be useful, although I imagine most looking at this repo are thinking about developing their own apiserver and are using this as a template to get started. So they will want to create their own images. At least I know, from my perspective, I wanted to know how to get setup, build and deploy. I then immediately wanted to make changes and add my own resource etc to see how that worked.

Reply:
The DOCKERHUB_USER must be replaced I agree. I meant the other changes, especially the `--requestheader-*` lines. I am proposing to either change the ones we have or if that conflicts with another use of them, let's create a different set of artifacts.

@deads2k can you comment where we use these artifacts now? Do we at all?

Reply:
@deads2k @sttts Yes agree, it would be better if we could add the required configuration to the existing artifacts

Reply:
> @deads2k can you comment where we use these artifacts now? Do we at all?

They can be auto-detected in the cluster if the user is bound with a role like this: https://github.com/openshift/kubernetes-namespace-reservation/blob/master/artifacts/install/rbac-template.yaml#L38-L50

Reply:
Yes from the configmap. But the issue (at least at the time of writing the walkthrough) was these setting were not present in minikube.